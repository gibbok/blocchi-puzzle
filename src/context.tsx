import React, { Provider } from 'react';
import { BrowserInfo } from './game/types';

export type AppContextType = Readonly<{
  browserInfo?: BrowserInfo;
  repeat: boolean;
  setRepeat: (repeat: boolean) => void;
}>;

export const AppContext = React.createContext<AppContextType | undefined>(undefined);

export function createCtx<ContextType>(): Readonly<
  [() => ContextType, Provider<ContextType | undefined>]
> {
  const context = React.createContext<ContextType | undefined>(undefined);

  function useContext() {
    const c = React.useContext(context);
    if (!c) {
      throw new Error('useContext must be inside a Provider with a value');
    }
    return c;
  }
  return [useContext, context.Provider] as const;
}

export const [useAppContextConsumer, AppContextProvider] = createCtx<AppContextType>();
