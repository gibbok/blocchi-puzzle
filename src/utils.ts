import { Tetro, Board } from './types';

export const logger = (x: Tetro | Board) => {
  console.log(x.reduce((acc, v) => `${acc} ${v.concat()} \n`, ''));
};
