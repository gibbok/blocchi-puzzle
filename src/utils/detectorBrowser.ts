import { detect } from 'detect-browser';

const getPathRelease = (version: string) => version.split('.')[0];

export const detectorBrowser = (): boolean => {
  const browser = detect();
  if (browser && browser.name && browser.version) {
    const name = browser.name;
    const version = browser.version;
    const path = Number(getPathRelease(version));

    switch (name) {
      case 'chrome':
      case 'firefox':
        return path > 80;
      case 'safari':
        return path > 13;
      case 'opera':
        return path > 62;
      default:
        return false;
    }
  } else {
    return false;
  }
};
