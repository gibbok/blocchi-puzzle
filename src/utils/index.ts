export * from './data.support.test';
export * from './logger';
