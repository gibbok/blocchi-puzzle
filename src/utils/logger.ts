import { Tetro, Board } from '../game/types';

export const logger = (x: Tetro | Board): void => {
  console.log(x.reduce((acc, v) => `${acc} ${v.concat()} \n`, ''));
};
