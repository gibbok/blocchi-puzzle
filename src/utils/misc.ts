import { IO, io } from 'fp-ts/lib/IO';
import { randomInt } from 'fp-ts/lib/Random';

export const getRandomValueFromStringEnum = <T>(strEnum: T): IO<T[keyof T]> => {
  const rndInt = randomInt(0, Object.keys(strEnum).length - 1)();
  const rndValue = Object.keys(strEnum)[rndInt] as keyof T;
  return io.of(strEnum[rndValue]);
};
