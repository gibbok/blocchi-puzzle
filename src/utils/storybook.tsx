import React from 'react';
import GlobalStyle from '../components/GlobalStyle';
import styled from 'styled-components';
import { wood } from '../assets/images';

/**
 * Decorate a story with glocal styles.
 * @param story Story
 */
export const globalStylesDecorator = (story: () => unknown): JSX.Element => (
  <React.Fragment>
    <GlobalStyle />
    {story()}
  </React.Fragment>
);

const Bkg = styled.div`
  background-image: url(${wood});
  width: 100vw;
  height: 100vh;
`;

/**
 * Decorate a story with a page background.
 * @param story Story
 */
export const pageBackground = (story: () => React.ReactNode): JSX.Element => <Bkg>{story()}</Bkg>;
