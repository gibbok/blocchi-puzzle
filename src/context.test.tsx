import React from 'react';
import { useAppContextConsumer, AppContextProvider } from './context';
import { mockContext } from './utils';
import renderer from 'react-test-renderer';

function TestComp(): JSX.Element {
  const { repeat, browserInfo, setRepeat } = useAppContextConsumer();
  return (
    <div>
      {JSON.stringify({
        repeat,
        browserInfo,
      })}
    </div>
  );
}

describe('context', () => {
  it('should use AppContextProvider', () => {
    const context = mockContext();
    const tree = renderer
      .create(
        <AppContextProvider value={context}>
          <TestComp />
        </AppContextProvider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should throw an exception if not used with an app context consumer', () => {
    let exception;
    try {
      const tree = renderer.create(<TestComp />);
      tree.toJSON;
    } catch (err) {
      exception = err;
    }

    expect(exception.message).toBe('useContext must be inside a Provider with a value');
  });
});
