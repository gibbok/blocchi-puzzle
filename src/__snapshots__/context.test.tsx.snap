// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context should use AppContextProvider 1`] = `
<div>
  {"repeat":false,"browserInfo":{"name":"chrome","version":"83"}}
</div>
`;
